contracts/modules/Checkpoint/Voting/APLCR/AdvancedPLCRVotingCheckpoint.sol

```diff
@@ -0,0 +1,1004 @@
+    {
+        createCustomStatutoryBallot(
+            _name,
+            _startTime,
+            _commitDuration,
+            _revealDuration,
+            _proposalTitle,
+            _details,
+            _choices,
+            _noOfChoices,
+            _checkpointId
+        );
+        _addExemptedAddresses(_exemptedAddresses, ballots.length - 1);
+    }
+
+    /**
+     * @notice Used to commit the vote
+     * @param _ballotId Given ballot Id
+     * @param _secretVote It is secret hash value (hashed offchain)
+     */
+    function commitVote(uint256 _ballotId, bytes32 _secretVote) external { 
+        // Check for the ballots array out of bound
+        _checkIndexOutOfBound(_ballotId);
+        require(_secretVote != bytes32(0), "Invalid vote");
+        // Check for the valid stage. Whether that ballot is in the COMMIT state or not.
+        _checkValidStage(_ballotId, Stage.COMMIT);
+        // Check whether the msg.sender is allowed to vote for a given ballotId or not. 
+        require(isVoterAllowed(_ballotId, msg.sender), "Invalid voter");
+        // validate the storage values
+        Ballot storage ballot = ballots[_ballotId];
+        require(ballot.voteDetails[msg.sender].secretVote == bytes32(0), "Already voted");
+        require(!ballot.isCancelled, "Cancelled ballot");
+        // Get the balance of the voter (i.e `msg.sender`) at the checkpoint on which ballot was created.
+        uint256 weight = securityToken.balanceOfAt(msg.sender, ballot.checkpointId);
+        require(weight > 0, "Zero weight is not allowed");
+        // Update the storage value.
+        ballot.voteDetails[msg.sender] = Vote(_secretVote);
+        emit VoteCommit(msg.sender, weight, _ballotId, _secretVote);
+    }
+
+    /**
+     * @notice Used to reveal the vote
+     * @param _ballotId Given ballot Id
+     * @param _choices Choices opt by the voter. 
+     * @dev Its length should be equal to the aggregation of ∑(proposalᵢ * noOfChoicesᵢ) where i = 0 ...totalProposal-1
+     * @param _salt used salt for hashing (unique for each user)
+     */
+    function revealVote(uint256 _ballotId, uint256[] calldata _choices, uint256 _salt) external {
+        // Check for the ballots array out of bound
+        _checkIndexOutOfBound(_ballotId);
+        // Check for the valid stage. Whether that ballot is in the REVEAL state or not.
+        _checkValidStage(_ballotId, Stage.REVEAL);
+        Ballot storage ballot = ballots[_ballotId];
+        // validate the storage values 
+        require(!ballot.isCancelled, "Cancelled ballot");
+        require(ballot.voteDetails[msg.sender].secretVote != bytes32(0), "Secret vote not available");
+        uint256 choiceCount = 0;
+        uint256 i;
+        for (i = 0; i < ballot.totalProposals; i++) {
+            uint256 _noOfChoice = ballot.proposals[i].noOfChoices;
+            _noOfChoice = _noOfChoice == 0 ? 3 : _noOfChoice;
```

Comment:
why is 3 the default?

Reply:
3 is used when your proposal is NAY/YAY type so when you will create the proposal with NAY/YAY type you will just pass no. of choices = 0 so the contract will automatically assign 3 choices to it.

Reply:
Maybe better to store 3 as default (rather than 0) when setting `noOfChoice` rather than here?

Reply:
If I store 3 as default then there is no way to dApp developers to know that proposal is NAY/YAY type. if there are 3 proposals with choices 3, 0, 3 but the smart contract will return 3, 3, 3 then during parsing the event data `_choices` it will be very hard to know which proposal is NAY/YAY type whether first or second or third. 
One thing I can do is create a constant variable with 3 as value and use that everywhere for better understanding.